numba_dppy/dppy_array_type.py

```diff
@@ -0,0 +1,95 @@
+# Copyright 2021 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from numba.core.types.npytypes import Array
+from numba.core import types
+from numba.core.datamodel.models import StructModel
+import numpy as np
+
+
+class DPPYArray(Array):
+    """
+    Type class for DPPY arrays.
+    """
+
+    def __init__(
+        self,
+        dtype,
+        ndim,
+        layout,
+        py_type=np.ndarray,
+        readonly=False,
+        name=None,
+        aligned=True,
+        addrspace=None,
+    ):
+        self.addrspace = addrspace
+        super(DPPYArray, self).__init__(
+            dtype,
+            ndim,
+            layout,
+            py_type=py_type,
+            readonly=readonly,
+            name=name,
+            aligned=aligned,
+        )
+
+    def copy(self, dtype=None, ndim=None, layout=None, readonly=None, addrspace=None):
+        if dtype is None:
+            dtype = self.dtype
+        if ndim is None:
+            ndim = self.ndim
+        if layout is None:
+            layout = self.layout
+        if readonly is None:
+            readonly = not self.mutable
+        if addrspace is None:
+            addrspace = self.addrspace
+        return DPPYArray(
+            dtype=dtype,
+            ndim=ndim,
+            layout=layout,
+            readonly=readonly,
+            aligned=self.aligned,
+            addrspace=addrspace,
+        )
+
+    @property
+    def key(self):
+        return (
+            self.dtype,
+            self.ndim,
+            self.layout,
+            self.mutable,
+            self.aligned,
+            self.addrspace,
+        )
+
+    def is_precise(self):
+        return self.dtype.is_precise()
+
+
+class DPPYArrayModel(StructModel):
+    def __init__(self, dmm, fe_type):
+        ndim = fe_type.ndim
+        members = [
+            ("meminfo", types.MemInfoPointer(fe_type.dtype)),
```

Comment:
Couple of questions @reazulhoque @DrTodd13 @PokhodenkoSA:

1. Do we need the `meminfo` and `parent` attributes? We are not using it in any place?
2. Will it be possible to add an address space specifier to the two attributes based on our `GenericPointerModel`?

It will be useful for my work on removing the kernel wrapper function. I made the needed changes so that address space casting of kernel arguments is done during the `decode_arguments` phase at the time of lowering. However, the next step of unflattening the array arguments into a struct fails because the data model does not match.

Reply:
Moved to https://github.com/IntelPython/numba-dppy/issues/364.
Could be implemented in separate PR.